composition-js/src/__tests__/compose.test.ts

```diff
@@ -4219,4 +4219,272 @@ describe('composition', () => {
+        typeDefs: gql`
+          enum ScopedEnum @requiresScopes(scopes: ["enum"]) {
+            A
+            B
+          }
+        `,
+        name: 'on-enum',
+      };
+
+      const onRootField = {
+        typeDefs: gql`
+          type Query {
+            scopedRootField: Int! @requiresScopes(scopes: ["rootField"])
+          }
+        `,
+        name: 'on-root-field',
+      };
+
+      const onObjectField = {
+        typeDefs: gql`
+          type Query {
+            objectWithField: ObjectWithScopedField!
+          }
+
+          type ObjectWithScopedField {
+            field: Int! @requiresScopes(scopes: ["objectField"])
+          }
+        `,
+        name: 'on-object-field',
+      };
+
+      const onEntityField = {
+        typeDefs: gql`
+          type Query {
+            entityWithField: EntityWithScopedField!
+          }
+
+          type EntityWithScopedField @key(fields: "id") {
+            id: ID!
+            field: Int! @requiresScopes(scopes: ["entityField"])
+          }
+        `,
+        name: 'on-entity-field',
+      };
+
+      const result = composeAsFed2Subgraphs([
+        onObject,
+        onInterface,
+        onInterfaceObject,
+        onScalar,
+        onEnum,
+        onRootField,
+        onObjectField,
+        onEntityField,
+      ]);
+      assertCompositionSuccess(result);
+
+      const scopedElements = [
+        "ScopedObject",
+        "ScopedInterface",
+        "ScopedInterfaceObject",
+        "ScopedScalar",
+        "ScopedEnum",
+        "Query.scopedRootField",
+        "ObjectWithScopedField.field",
+        "EntityWithScopedField.field",
+      ];
+
+      for (const element of scopedElements) {
+        expect(
+          result.schema
+            .elementByCoordinate(element)
+            ?.hasAppliedDirective("requiresScopes")
+        ).toBeTruthy();
+      }
+    });
+
+    it('applies @requiresScopes on types as long as it is used once', () => {
+      const a1 = {
+        typeDefs: gql`
+          type Query {
+            a: A
+          }
+          type A @key(fields: "id") @requiresScopes(scopes: ["a"]) {
+            id: String!
+            a1: String
+          }
+        `,
+        name: 'a1',
+      };
+      const a2 = {
+        typeDefs: gql`
+          type A @key(fields: "id") {
+            id: String!
+            a2: String
+          }
+        `,
+        name: 'a2',
+      };
+
+      // checking composition in either order (not sure if this is necessary but
+      // it's not hurting anything)
+      const result1 = composeAsFed2Subgraphs([a1, a2]);
+      const result2 = composeAsFed2Subgraphs([a2, a1]);
+      assertCompositionSuccess(result1);
+      assertCompositionSuccess(result2);
+
+      expect(result1.schema.type('A')?.hasAppliedDirective('requiresScopes')).toBeTruthy();
+      expect(result2.schema.type('A')?.hasAppliedDirective('requiresScopes')).toBeTruthy();
+    });
+
+    it('merges @requiresScopes lists', () => {
```

Comment:
For the sake of completeness, might we worth having a test that merge scopes that intersects (say `["a", "b"]` on the one side and `["b", "c"]` on the other).

Reply:
Good call, added